### PR TITLE
Add swift-crypto dependency

### PR DESCRIPTION
**Pull request by yim-lee**
Motivation:
Upcoming package collections feature will require swift-crypto dependency.

Modifications:
- Add swift-crypto dependency
- Bump minimum macOS version to 10.15, for it's required by swift-crypto

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
Does this mean we need to add `CMakeLists` files to swift-crypto?

```
11:03:28 -- Check for working Swift compiler: /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/build/buildbot_incremental/toolchain-linux-x86_64/usr/bin/swiftc
11:03:29 -- Check for working Swift compiler: /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/build/buildbot_incremental/toolchain-linux-x86_64/usr/bin/swiftc -- works
11:03:29 CMake Error at CMakeLists.txt:50 (find_package):
11:03:29   Could not find a package configuration file provided by "Crypto" with any
11:03:29   of the following names:
11:03:29 
11:03:29     CryptoConfig.cmake
11:03:29     crypto-config.cmake
11:03:29 
11:03:29   Add the installation prefix of "Crypto" to CMAKE_PREFIX_PATH or set
11:03:29   "Crypto_DIR" to a directory containing one of the above files.  If "Crypto"
11:03:29   provides a separate development package or SDK, be sure it has been
11:03:29   installed.
```

**Comment by tomerd** [comment]
> Right, it looks as if swift-crypto isn't able to build with CMake. But is it needed for bootstrapping? CMake is only used in order to get a swift-build that can build itself

I dont know that that is true for Windows (tho would be nice to have minimal pre step) @compnerd ?

**Comment by buttaface** [comment]
Unless you plan on using these new crypto features when building SPM with itself, which shouldn't be necessary since `build-script` should have already checked out SPM's dependencies, I think it's fine to not add this swift-crypto dependency for the CMake bootstrap and avoid adding a CMake config for swift-crypto.

**Comment by compnerd** [comment]
Windows definitely relies on the CMake build to get a fully functional SPM.  The manifest execution is a problem - there is no concept of rpath, and static linking for Swift modules is unsupported.  We currently are reliant on CMake for the complete build.

**Comment by buttaface** [comment]
Alright, I didn't know you weren't building SPM with SPM on Windows and are using the CMake-configured build of SPM instead on that platform.

**Comment by neonichu** [comment]
@yim-lee BTW, if we land https://github.com/apple/swift/pull/35507, we could use cross-repository testing together with the swift-crypto PR -- might be worth it to get a fully end-to-end test on CI before we land the changes in swift-crypto? Then maybe pinning a version in `update-checkout` could be a follow-up PR.

File: Sources/PackageCollections/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,7 @@ target_link_libraries(PackageCollections PUBLIC
   TSCBasic
   TSCUtility
   Basics
+  Crypto
```
Comment (yim-lee) [inline Sources/PackageCollections/CMakeLists.txt:35]: @compnerd I try running the CMake build in local docker via `./Utilities/bootstrap build` with swift-crypto `main`, but for some reason `Crypto` is not found:

```
--- bootstrap: note: Building SwiftPM (with CMake)
[1/1] Linking Swift shared library lib/libPackageCollections.so
FAILED: lib/libPackageCollections.so Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/Collection.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/CVE.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/License.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/Package.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/Search.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Model/TargetListResult.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Providers/GitHubPackageMetadataProvider.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Providers/JSONPackageCollectionProvider.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Providers/PackageCollectionProvider.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Providers/PackageMetadataProvider.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Storage/FilePackageCollectionsSourcesStorage.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Storage/PackageCollectionsSourcesStorage.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Storage/PackageCollectionsStorage.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Storage/SQLitePackageCollectionsStorage.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Storage/Trie.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/API.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/PackageCollections.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/PackageCollections+Configuration.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/PackageCollections+Storage.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/PackageCollections+Validation.swift.o Sources/PackageCollections/CMakeFiles/PackageCollections.dir/Utility.swift.o swift/PackageCollections.swiftmodule 
: && /usr/bin/swiftc -output-file-map Sources/PackageCollections/CMakeFiles/PackageCollections.dir/output-file-map.json -incremental -j 6 -emit-library -o lib/libPackageCollections.so -module-name PackageCollections -module-link-name PackageCollections -emit-module -emit-module-path swift/PackageCollections.swiftmodule -emit-dependencies -DPackageCollections_EXPORTS -module-cache-path "/code/swift-package-manager/.build/x86_64-unknown-linux-gnu/bootstrap/module-cache" -g -swift-version 5 -I swift -I ../tsc/swift ../../../Sources/PackageCollections/Model/Collection.swift ../../../Sources/PackageCollections/Model/CVE.swift ../../../Sources/PackageCollections/Model/License.swift ../../../Sources/PackageCollections/Model/Package.swift ../../../Sources/PackageCollections/Model/Search.swift ../../../Sources/PackageCollections/Model/TargetListResult.swift ../../../Sources/PackageCollections/Providers/GitHubPackageMetadataProvider.swift ../../../Sources/PackageCollections/Providers/JSONPackageCollectionProvider.swift ../../../Sources/PackageCollections/Providers/PackageCollectionProvider.swift ../../../Sources/PackageCollections/Providers/PackageMetadataProvider.swift ../../../Sources/PackageCollections/Storage/FilePackageCollectionsSourcesStorage.swift ../../../Sources/PackageCollections/Storage/PackageCollectionsSourcesStorage.swift ../../../Sources/PackageCollections/Storage/PackageCollectionsStorage.swift ../../../Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift ../../../Sources/PackageCollections/Storage/Trie.swift ../../../Sources/PackageCollections/API.swift ../../../Sources/PackageCollections/PackageCollections.swift ../../../Sources/PackageCollections/PackageCollections+Configuration.swift ../../../Sources/PackageCollections/PackageCollections+Storage.swift ../../../Sources/PackageCollections/PackageCollections+Validation.swift ../../../Sources/PackageCollections/Utility.swift  -Xlinker -soname -Xlinker libPackageCollections.so  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/bootstrap/lib  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/bootstrap/lib  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/bootstrap/lib  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/tsc/lib  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/tsc/lib  -L /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/tsc/lib -Xlinker -rpath -Xlinker /code/swift-package-manager/.build/x86_64-unknown-linux-gnu/bootstrap/lib:/code/swift-package-manager/.build/x86_64-unknown-linux-gnu/tsc/lib  -lCrypto  lib/libPackageModel.so  lib/libSourceControl.so  lib/libBasics.so  ../tsc/lib/libTSCUtility.so  ../tsc/lib/libTSCBasic.so  ../tsc/lib/libTSCLibc.so && :
/usr/bin/ld.gold: error: cannot find -lCrypto
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
/usr/bin/ninja
--- bootstrap: error: Command '['/usr/bin/ninja']' returned non-zero exit status 1.
```

Any ideas on why?

The `find_package(Crypto CONFIG REQUIRED)` call in root `CMakeLists.txt` seems to have completed successfully. 

`libCrypto.so` exists at the proper location I think: 

```
> ls -al .build/x86_64-unknown-linux-gnu/swift-crypto/lib/
total 27008
drwxr-xr-x   5 ylee  staff      160 Jan 27 23:01 .
drwxr-xr-x  14 ylee  staff      448 Jan 27 23:01 ..
-rw-r--r--@  1 ylee  staff  9094506 Jan 27 23:01 libCCryptoBoringSSL.a
-rw-r--r--@  1 ylee  staff    23238 Jan 27 23:01 libCCryptoBoringSSLShims.a
-rwxr-xr-x   1 ylee  staff  4324720 Jan 27 23:01 libCrypto.so
```

At least I don't see swift-crypto's `CMakeLists.txt`s doing something different from `TSC`, `ArgumentParser`, etc.

Comment (yim-lee) [reply Sources/PackageCollections/CMakeLists.txt:35]: Found a potential cause: I'm not consistent with naming `SWIFT_CRYPTO_EXPORTS` vs. `CRYPTO_EXPORTS`. Will look into that.

Comment (yim-lee) [reply Sources/PackageCollections/CMakeLists.txt:35]: Indeed that was the issue.

File: Sources/PackageCollections/Signing/PackageCollectionSignature.swift [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Crypto
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import Security
+#else
+import CCryptoBoringSSL
+#endif
```
Comment (yim-lee) [inline Sources/PackageCollections/Signing/PackageCollectionSignature.swift:16]: This is how swift-crypto will be used. We will need the `Security` framework as well. I added it so we can test everything together.

**Comment by yim-lee** [comment]
I tested locally that with this PR + https://github.com/apple/swift-crypto/pull/65 cmake build completes successfully on Mac and Linux (`swift:5.3-focal`).

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
Build error seems unrelated:

```
/home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/swift-driver/Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:938:86: error: 'SwiftDeps' is not a member type of class 'SwiftDriver.ModuleDependencyGraph'
20:05:02 extension BidirectionalMap where T1 == TypedVirtualPath, T2 == ModuleDependencyGraph.SwiftDeps {
20:05:02                                                                ~~~~~~~~~~~~~~~~~~~~~ ^
20:05:02 /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/swift-driver/Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:20:39: note: 'ModuleDependencyGraph' declared here
20:05:02 /*@_spi(Testing)*/ public final class ModuleDependencyGraph {
20:05:02                                       ^
20:05:02 /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-main/swift-driver/Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:923:7: error: cannot find 'sourceSwiftDepsMap' in scope
20:05:02       sourceSwiftDepsMap.matches(other.sourceSwiftDepsMap),
20:05:02       ^~~~~~~~~~~~~~~~~~
```

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
We've moved past the cmake build issues. The "Swift Test macOS Platform (smoke test)" failure is:

```
22:32:38 error: the library 'SKSwiftPMWorkspace' requires macos 10.10, but depends on the product 'SwiftPM-auto' which requires macos 10.15; consider changing the library 'SKSwiftPMWorkspace' to require macos 10.15 or later, or the product 'SwiftPM-auto' to require macos 10.10 or earlier.
```

@neonichu Is this the test to update? https://github.com/apple/sourcekit-lsp/blob/main/Tests/SKSwiftPMWorkspaceTests/SwiftPMWorkspaceTests.swift

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]

**Comment by yim-lee** [comment]
🥳 all green!

File: Sources/PackageCollections/Signing/PackageCollectionSignature.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/PackageCollections/Signing/PackageCollectionSignature.swift:16]: Great, this stub is really helpful to see how it gets used.

**Review by abertelrud — APPROVED** [review]
LGTM

**Review by neonichu — APPROVED** [review]

**Comment by yim-lee** [comment]
Testing https://github.com/apple/swift-crypto/pull/67

@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
🥳

**Pull request merged**